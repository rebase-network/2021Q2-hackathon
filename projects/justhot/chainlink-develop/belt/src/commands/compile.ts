import { Command, flags } from '@oclif/command'
import * as Parser from '@oclif/parser'
import { App } from '../services/config'

export default class Compile extends Command {
  static description =
    'Run various compilers and/or codegenners that target solidity smart contracts.'

  static examples = [
    `$ belt compile all

Creating directory at abi/v0.4...
Creating directory at abi/v0.5...
Creating directory at abi/v0.6...
Compiling 35 contracts...
...
...
Aggregator artifact saved!
AggregatorProxy artifact saved!
Chainlink artifact saved!
...`,
  ]

  static flags = {
    help: flags.help({ char: 'h' }),
    config: flags.string({
      char: 'c',
      default: 'app.config.json',
      description: 'Location of the configuration file',
    }),
  }

  static args: Parser.args.IArg[] = [
    {
      name: 'compiler',
      description:
        'Compile solidity smart contracts and output their artifacts',
      options: ['solc', 'ethers', 'truffle', 'all'],
    },
  ]

  async run() {
    const { flags, args, argv } = this.parse(Compile)
    if (argv.length === 0) {
      this._help()
    }

    try {
      const config = await import('../services/config')
      const conf = config.load(flags.config)
      const compilation =
        args.compiler === 'all'
          ? this.compileAll(conf)
          : this.compileSingle(args.compiler, conf)

      await compilation
    } catch (e) {
      this.error(e)
    }
  }

  private async compileSingle(compilerName: string, conf: App) {
    type Compilers = typeof import('../services/compilers')
    const compiler: Compilers[keyof Compilers] = await import(
      `../services/compilers/${compilerName}`
    )

    await compiler.compileAll(conf)
  }

  private async compileAll(conf: App) {
    const compilers = await import('../services/compilers')

    await compilers.solc.compileAll(conf)
    await Promise.all([
      compilers.truffle.compileAll(conf),
      compilers.ethers.compileAll(conf),
    ])
  }
}
