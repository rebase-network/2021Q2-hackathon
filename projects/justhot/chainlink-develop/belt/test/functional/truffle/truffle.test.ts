import { join } from 'path'
import { ls, rm } from 'shelljs'
import compile from '../../../src/commands/compile'

const TEST_PATH = 'test/functional/truffle/'
const TEST_FS_PATH = join(TEST_PATH, 'testfs')
const FIXTURES_PATH = join(TEST_PATH, 'fixtures')

describe('compileAll', () => {
  beforeEach(() => {
    rm('-r', TEST_FS_PATH)
  })

  it('should produce a truffle contract abstraction via a json artifact produced by sol-compiler', async () => {
    await compile.run([
      `--config=${join(FIXTURES_PATH, 'test.config.json')}`,
      'truffle',
    ])

    expect([...ls('-R', TEST_FS_PATH)]).toMatchSnapshot()
  })
})
