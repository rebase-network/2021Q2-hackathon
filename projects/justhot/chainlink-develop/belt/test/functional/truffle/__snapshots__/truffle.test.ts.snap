// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compileAll should produce a truffle contract abstraction via a json artifact produced by sol-compiler 1`] = `
Array [
  "truffle",
  "truffle/v0.4",
  "truffle/v0.4/Ownable.js",
  "truffle/v0.5",
  "truffle/v0.5/Ownable.js",
]
`;
